File: sdk/csharp/Orbit.Sdk/Accounts.cs
using System.Text.Json;

namespace Orbit.Sdk;

public sealed record Customer(string Id, string Username, string Email, bool Suspended, string CreatedAt);
public sealed record Account(Customer Customer, string ExpiresAt);
public sealed record OwnedLicence(string Id, string PolicyName, string State, string ExpiryMode,
    string? FirstUsedAt, string? ExpiresAt, long? DurationSeconds, int DeviceLimit, bool HwidLocked,
    bool OfflineAllowed, int OfflineSeconds, IReadOnlyDictionary<string, bool> Entitlements);
public sealed record OwnedLicences(IReadOnlyList<OwnedLicence> Items, string? NextCursor);

/// <summary>
/// Sensitive login proof held only in memory. Possession establishes neither
/// current session validity nor licensed access.
/// </summary>
public sealed class CustomerSessionProof
{
    private readonly string token;

    internal CustomerSessionProof(string token) { this.token = token; }

    /// <summary>
    /// Returns the sensitive Bearer authorization header value. Send it only to
    /// your own trusted HTTPS backend, which must verify the session online with
    /// Orbit. Never log or persist it, send it to arbitrary URLs, or forward it
    /// through redirects. This proof grants no licensed access.
    /// </summary>
    public string AuthorizationHeader() => "Bearer " + token;

    public override string ToString() => "Orbit customer session proof (redacted)";
}

/// <summary>Request input only. The client never retains passwords or raw licence keys.</summary>
public sealed class Registration(string licenceKey, string username, string email, string password)
{
    internal string LicenceKey { get; } = licenceKey;
    internal string Username { get; } = username;
    internal string Email { get; } = email;
    internal string Password { get; } = password;
    public override string ToString() => "Orbit registration (redacted)";
}

/// <summary>A scoped resend proof held only in memory. Not an account session or an access grant.</summary>
public sealed class PendingRegistration
{
    public bool Accepted => true;
    public string ExpiresAt { get; }
    internal string ResendCredential { get; }
    internal string ApplicationId { get; }
    internal string EnvironmentId { get; }
    internal PendingRegistration(string expiresAt, string resendCredential, OrbitConfig config)
    {
        ExpiresAt = expiresAt;
        ResendCredential = resendCredential;
        ApplicationId = config.ApplicationId;
        EnvironmentId = config.EnvironmentId;
    }
    public override string ToString() => "Orbit pending registration (redacted)";
}

internal sealed class CustomerSession(string token, Account account)
{
    internal string Token { get; } = token;
    internal Account Account { get; } = account;
}

public sealed partial class OrbitClient
{
    /// <summary>Local account metadata only; this grants no licensed access.</summary>
    public Account? Account()
    {
        lock (gate) { SyncStorage(); return session?.Account; }
    }

    /// <summary>
    /// Copies locally held login proof after synchronizing storage invalidation.
    /// This does not establish online validity or licensed access.
    /// </summary>
    public CustomerSessionProof CustomerSessionProof()
    {
        lock (gate)
        {
            SyncStorage();
            if (session == null) throw new OrbitException(OrbitError.ReauthenticationRequired);
            return new CustomerSessionProof(session.Token);
        }
    }

    public async Task<Account> LoginAsync(string username, string password, CancellationToken cancellationToken = default)
    {
        if (username.Length is < 1 or > 128 || password.Length > 256) throw new OrbitException(OrbitError.Configuration);
        var expected = Generation();
        await EnterSerialAsync(cancellationToken).ConfigureAwait(false);
        try
        {
            lock (gate)
            {
                CheckGeneration(expected);
                OrbitException.CheckCancellation(cancellationToken);
                Clear();
                InvalidateStorage();
                expected = generation;
            }
            try
            {
                var body = ScopeBody();
                body["username"] = username;
                body["password"] = password;
                var reply = await transport.PostAsync("/api/client/v1/sessions", body, false, cancellationToken).ConfigureAwait(false)
                    ?? throw JsonWire.Invalid();
                var customer = ReadCustomer(JsonWire.Field(reply, "customer"));
                var token = JsonWire.String(reply, "session");
                var expires = JsonWire.String(reply, "expires_at");
                _ = JsonWire.Timestamp(expires);
                if (!JsonWire.Bearer(token)) throw JsonWire.Invalid();
                var account = new Account(customer, expires);
                lock (gate)
                {
                    CheckGeneration(expected);
                    OrbitException.CheckCancellation(cancellationToken);
                    session = new CustomerSession(token, account);
                    return account;
                }
            }
            catch (OrbitException error) { throw FinishAccountError(error, expected, cancellationToken); }
        }
        finally { serial.Release(); }
    }

    public Task<Snapshot> ActivateAccountAsync(string licenceId, string idempotencyKey, CancellationToken cancellationToken = default) =>
        ActivateAccountWithPreviousAsync(licenceId, null, idempotencyKey, cancellationToken);

    public Task<Snapshot> ActivateAccountWithPreviousAsync(string licenceId, string? previousCredential, string idempotencyKey,
        CancellationToken cancellationToken = default) => ActivateAsAsync(licenceId, true, previousCredential, idempotencyKey, cancellationToken);

    /// <summary>Clears local state synchronously. Await the returned task to confirm remote session revocation.</summary>
    public Task LogoutAccountAsync(CancellationToken cancellationToken = default)
    {
        CustomerSession? saved;
        long expected;
        lock (gate)
        {
            SyncStorage();
            saved = session;
            Clear();
            InvalidateStorage();
            expected = generation;
        }
        return saved == null ? Task.CompletedTask : FinishLogoutAsync(saved, expected, cancellationToken);
    }

    private async Task FinishLogoutAsync(CustomerSession saved, long expected, CancellationToken cancellationToken)
    {
        try
        {
            var reply = await transport.DeleteAsync(ScopePath("/api/client/v1/sessions/current"), saved.Token, cancellationToken).ConfigureAwait(false);
            lock (gate) { CheckGeneration(expected); OrbitException.CheckCancellation(cancellationToken); }
            if (reply != null) throw JsonWire.Invalid();
        }
        catch (OrbitException)
        {
            lock (gate) CheckGeneration(expected);
            throw;
        }
    }

    public async Task<OwnedLicences> OwnedLicencesAsync(string? after = null, CancellationToken cancellationToken = default)
    {
        if (after != null && !JsonWire.Opaque(after)) throw new OrbitException(OrbitError.Configuration);
        var expected = Generation();
        await EnterSerialAsync(cancellationToken).ConfigureAwait(false);
        try
        {
            var current = GetSession(expected);
            try
            {
                var reply = await transport.GetAsync(ScopePath("/api/client/v1/licences", after), cancellationToken, current.Token).ConfigureAwait(false)
                    ?? throw JsonWire.Invalid();
                var entries = JsonWire.Field(reply, "items");
                if (entries.ValueKind != JsonValueKind.Array || entries.GetArrayLength() > 50) throw JsonWire.Invalid();
                var items = entries.EnumerateArray().Select(ReadLicence).ToArray();
                var cursor = JsonWire.OptionalString(reply, "next_cursor");
                if (cursor != null && !JsonWire.Opaque(cursor)) throw JsonWire.Invalid();
                return FinishAccount(new OwnedLicences(Array.AsReadOnly(items), cursor), expected, cancellationToken);
            }
            catch (OrbitException error) { throw FinishAccountError(error, expected, cancellationToken); }
        }
        finally { serial.Release(); }
    }

    public Task<OwnedLicence> ClaimLicenceAsync(string licenceKey, string idempotencyKey, CancellationToken cancellationToken = default)
    {
        if (licenceKey.Length is < 1 or > 256 || !JsonWire.OperationId(idempotencyKey)) throw new OrbitException(OrbitError.Configuration);
        var body = ScopeBody();
        body["licence_key"] = licenceKey;
        body["idempotency_key"] = idempotencyKey;
        return AccountPostAsync("/api/client/v1/licence-claims", body, true, ReadLicence, cancellationToken);
    }

    public async Task RequestEmailChangeAsync(string password, string email, CancellationToken cancellationToken = default)
    {
        if (password.Length > 256 || email.Length > 254) throw new OrbitException(OrbitError.Configuration);
        var body = ScopeBody();
        body["password"] = password;
        body["email"] = email;
        _ = await AccountPostAsync("/api/client/v1/email-changes", body, false, Accepted, cancellationToken).ConfigureAwait(false);
    }

    public async Task<PendingRegistration> RegisterAsync(Registration registration, CancellationToken cancellationToken = default)
    {
        if (registration.LicenceKey.Length is < 1 or > 256 || registration.Username.Length > 128 ||
            registration.Email.Length > 254 || registration.Password.Length > 256 ||
            registration.Password.EnumerateRunes().Count() < 8) throw new OrbitException(OrbitError.Configuration);
        var body = ScopeBody();
        body["licence_key"] = registration.LicenceKey;
        body["username"] = registration.Username;
        body["email"] = registration.Email;
        body["password"] = registration.Password;
        var reply = await transport.PostAsync("/api/client/v1/registrations", body, false, cancellationToken).ConfigureAwait(false)
            ?? throw JsonWire.Invalid();
        _ = Accepted(reply);
        var proof = JsonWire.String(reply, "resend_credential");
        var expiry = JsonWire.String(reply, "expires_at");
        _ = JsonWire.Timestamp(expiry);
        if (!JsonWire.Bearer(proof)) throw JsonWire.Invalid();
        return new PendingRegistration(expiry, proof, config);
    }

    public async Task ResendRegistrationAsync(PendingRegistration pending, CancellationToken cancellationToken = default)
    {
        if (pending.ApplicationId != config.ApplicationId || pending.EnvironmentId != config.EnvironmentId)
            throw new OrbitException(OrbitError.Configuration);
        var body = ScopeBody();
        body["resend_credential"] = pending.ResendCredential;
        _ = Accepted(await transport.PostAsync("/api/client/v1/registrations/resend", body, false, cancellationToken).ConfigureAwait(false)
            ?? throw JsonWire.Invalid());
    }

    public async Task RequestPasswordRecoveryAsync(string email, CancellationToken cancellationToken = default)
    {
        if (email.Length > 254) throw new OrbitException(OrbitError.Configuration);
        var body = ScopeBody();
        body["email"] = email;
        _ = Accepted(await transport.PostAsync("/api/client/v1/password-recovery", body, false, cancellationToken).ConfigureAwait(false)
            ?? throw JsonWire.Invalid());
    }

    private CustomerSession GetSession(long expected)
    {
        lock (gate)
        {
            CheckGeneration(expected);
            return session ?? throw new OrbitException(OrbitError.ReauthenticationRequired);
        }
    }

    private async Task<T> AccountPostAsync<T>(string path, Dictionary<string, object?> body, bool retrySafe,
        Func<JsonElement, T> decode, CancellationToken cancellationToken)
    {
        var expected = Generation();
        await EnterSerialAsync(cancellationToken).ConfigureAwait(false);
        try
        {
            body["customer_session"] = GetSession(expected).Token;
            try
            {
                var response = await transport.PostAsync(path, body, retrySafe, cancellationToken).ConfigureAwait(false)
                    ?? throw JsonWire.Invalid();
                return FinishAccount(decode(response), expected, cancellationToken);
            }
            catch (OrbitException error) { throw FinishAccountError(error, expected, cancellationToken); }
        }
        finally { serial.Release(); }
    }

    private T FinishAccount<T>(T result, long expected, CancellationToken cancellationToken)
    {
        lock (gate)
        {
            CheckGeneration(expected);
            OrbitException.CheckCancellation(cancellationToken);
            return result;
        }
    }

    private OrbitException FinishAccountError(OrbitException error, long expected, CancellationToken cancellationToken)
    {
        lock (gate)
        {
            CheckGeneration(expected);
            OrbitException.CheckCancellation(cancellationToken);
            // Login-session expiry has a separate lifetime from activation access.
            if (error.Error is not (OrbitError.Transient or OrbitError.Cancelled) &&
                !(error.Error == OrbitError.Denied && error.Code == "session_expired"))
            {
                Clear();
                InvalidateStorage();
            }
            return error;
        }
    }

    private static bool Accepted(JsonElement reply) => JsonWire.Boolean(reply, "accepted") ? true : throw JsonWire.Invalid();

    private static Customer ReadCustomer(JsonElement value)
    {
        var customer = new Customer(JsonWire.String(value, "id"), JsonWire.String(value, "username"),
            JsonWire.String(value, "email"), JsonWire.Boolean(value, "suspended"), JsonWire.String(value, "created_at"));
        if (!JsonWire.Opaque(customer.Id) || customer.Suspended || customer.Username.Length is < 3 or > 32 ||
            !customer.Username.All(c => c is >= 'a' and <= 'z' or >= '0' and <= '9' or '_') || customer.Email.Length is < 1 or > 254)
            throw JsonWire.Invalid();
        _ = JsonWire.Timestamp(customer.CreatedAt);
        return customer;
    }

    private static OwnedLicence ReadLicence(JsonElement value)
    {
        var id = JsonWire.String(value, "id");
        var policy = JsonWire.String(value, "policy_name");
        var deviceLimit = JsonWire.Integer(value, "device_limit");
        var offlineSeconds = JsonWire.Integer(value, "offline_seconds");
        if (!JsonWire.Opaque(id) || policy.EnumerateRunes().Count() > 80 || deviceLimit is < 1 or > 100 || offlineSeconds is < 0 or > 86400)
            throw JsonWire.Invalid();
        var firstUsed = JsonWire.OptionalString(value, "first_used_at");
        var expires = JsonWire.OptionalString(value, "expires_at");
        if (firstUsed != null) _ = JsonWire.Timestamp(firstUsed);
        if (expires != null) _ = JsonWire.Timestamp(expires);
        return new OwnedLicence(id, policy, JsonWire.String(value, "state"), JsonWire.String(value, "expiry_mode"),
            firstUsed, expires, JsonWire.OptionalInteger(value, "duration_seconds"), (int)deviceLimit,
            JsonWire.Boolean(value, "hwid_locked"), JsonWire.Boolean(value, "offline_allowed"), (int)offlineSeconds,
            JsonWire.Entitlements(JsonWire.Field(value, "entitlements")));
    }
}

File: sdk/go/accounts.go
package orbit

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"strings"
	"unicode/utf8"
)

// Customer and Account contain display metadata; login grants no feature access.
type Customer struct {
	ID        string `json:"id"`
	Username  string `json:"username"`
	Email     string `json:"email"`
	Suspended bool   `json:"suspended"`
	CreatedAt string `json:"created_at"`
}
type Account struct {
	Customer  Customer
	ExpiresAt string
}

// CustomerSessionProof holds sensitive login proof only in memory. Possession
// establishes neither current session validity nor licensed access.
type CustomerSessionProof struct {
	token string
}

// AuthorizationHeader returns the sensitive Bearer authorization header value.
// Send it only to your own trusted HTTPS backend, which must verify the session
// online with Orbit. Never log or persist it, send it to arbitrary URLs, or
// forward it through redirects. This proof grants no licensed access.
func (p CustomerSessionProof) AuthorizationHeader() string {
	return "Bearer " + p.token
}

func (CustomerSessionProof) Format(f fmt.State, _ rune) {
	_, _ = f.Write([]byte("[Orbit customer session proof redacted]"))
}
func (CustomerSessionProof) MarshalJSON() ([]byte, error) { return nil, ErrConfiguration }

type accountSession struct {
	token   string
	account Account
}
type loginReply struct {
	Customer  Customer `json:"customer"`
	Session   string   `json:"session"`
	ExpiresAt string   `json:"expires_at"`
}

type OwnedLicence struct {
	ID              string          `json:"id"`
	PolicyName      string          `json:"policy_name"`
	State           string          `json:"state"`
	ExpiryMode      string          `json:"expiry_mode"`
	FirstUsedAt     *string         `json:"first_used_at"`
	ExpiresAt       *string         `json:"expires_at"`
	DurationSeconds *int64          `json:"duration_seconds"`
	DeviceLimit     int32           `json:"device_limit"`
	HWIDLocked      bool            `json:"hwid_locked"`
	OfflineAllowed  bool            `json:"offline_allowed"`
	OfflineSeconds  int32           `json:"offline_seconds"`
	Entitlements    map[string]bool `json:"entitlements"`
}
type OwnedLicences struct {
	Items      []OwnedLicence `json:"items"`
	NextCursor *string        `json:"next_cursor"`
}

// Registration is sent once and never retained by the client. Avoid logging it.
type Registration struct {
	LicenceKey string
	Username   string
	Email      string
	Password   string
}

func (Registration) Format(f fmt.State, _ rune) {
	_, _ = f.Write([]byte("[Orbit registration redacted]"))
}
func (Registration) MarshalJSON() ([]byte, error) { return nil, ErrConfiguration }

// PendingRegistration keeps the scoped resend proof private in memory.
type PendingRegistration struct {
	Accepted         bool
	ExpiresAt        string
	resendCredential string
	applicationID    string
	environmentID    string
}

func (PendingRegistration) Format(f fmt.State, _ rune) {
	_, _ = f.Write([]byte("[Orbit pending registration redacted]"))
}

type registrationReply struct {
	Accepted         bool   `json:"accepted"`
	ExpiresAt        string `json:"expires_at"`
	ResendCredential string `json:"resend_credential"`
}
type acceptedReply struct {
	Accepted bool `json:"accepted"`
}

func (c *Client) Account() (*Account, error) {
	c.mu.Lock()
	defer c.mu.Unlock()
	if err := c.syncStorageLocked(); err != nil {
		return nil, err
	}
	if c.state.account == nil {
		return nil, nil
	}
	copied := c.state.account.account
	return &copied, nil
}

// CustomerSessionProof copies locally held login proof after synchronizing
// storage invalidation. It does not establish online validity or licensed access.
func (c *Client) CustomerSessionProof() (*CustomerSessionProof, error) {
	c.mu.Lock()
	defer c.mu.Unlock()
	if err := c.syncStorageLocked(); err != nil {
		return nil, err
	}
	if c.state.account == nil {
		return nil, ErrReauthenticationRequired
	}
	return &CustomerSessionProof{token: strings.Clone(c.state.account.token)}, nil
}

func (c *Client) Login(ctx context.Context, username, password string) (Account, error) {
	if username == "" || len(username) > 128 || len(password) > 256 || !utf8.ValidString(password) {
		return Account{}, ErrConfiguration
	}
	generation, err := c.generation()
	if err != nil {
		return Account{}, err
	}
	if err := c.lockOperation(ctx, generation); err != nil {
		return Account{}, err
	}
	defer c.unlockOperation()
	c.mu.Lock()
	if err := c.syncStorageLocked(); err != nil {
		c.mu.Unlock()
		return Account{}, err
	}
	if c.state.generation != generation {
		c.mu.Unlock()
		return Account{}, ErrStaleResponse
	}
	clearState(&c.state)
	if err := c.invalidateLocked(); err != nil {
		c.mu.Unlock()
		return Account{}, err
	}
	generation = c.state.generation
	c.mu.Unlock()
	data, err := c.transport.Post(ctx, clientPrefix+"sessions", c.scopeBody(map[string]any{"username": username, "password": password}), false)
	var reply loginReply
	if err == nil {
		err = decodeJSON(data, &reply)
	}
	if err == nil {
		err = checkLogin(reply)
	}
	if err := c.finishAccount(ctx, generation, err); err != nil {
		return Account{}, err
	}
	account := Account{Customer: reply.Customer, ExpiresAt: reply.ExpiresAt}
	c.mu.Lock()
	defer c.mu.Unlock()
	if err := c.syncStorageLocked(); err != nil {
		return Account{}, err
	}
	if c.state.generation != generation {
		return Account{}, ErrStaleResponse
	}
	if ctx.Err() != nil {
		return Account{}, ErrCancelled
	}
	c.state.account = &accountSession{token: reply.Session, account: account}
	return account, nil
}

func (c *Client) ActivateAccount(ctx context.Context, licenceID, idempotencyKey string) (Snapshot, error) {
	return c.ActivateAccountWithPrevious(ctx, licenceID, "", idempotencyKey)
}
func (c *Client) ActivateAccountWithPrevious(ctx context.Context, licenceID, previousCredential, idempotencyKey string) (Snapshot, error) {
	if !opaque(licenceID) {
		return Snapshot{}, ErrConfiguration
	}
	return c.activate(ctx, "", licenceID, previousCredential, idempotencyKey)
}

// LogoutAccount clears local access before networking, even when ctx is already
// cancelled. A nil result confirms remote revocation when a session was present.
func (c *Client) LogoutAccount(ctx context.Context) error {
	c.mu.Lock()
	if err := c.syncStorageLocked(); err != nil {
		c.mu.Unlock()
		return err
	}
	session := c.state.account
	clearState(&c.state)
	if err := c.invalidateLocked(); err != nil {
		c.mu.Unlock()
		return err
	}
	generation := c.state.generation
	c.mu.Unlock()
	if session == nil {
		return nil
	}
	data, err := c.transport.DeleteBearer(ctx, c.accountPath(clientPrefix+"sessions/current", ""), session.token)
	if generationErr := c.checkGeneration(generation); generationErr != nil {
		return generationErr
	}
	if err != nil {
		return err
	}
	if data != nil {
		return ErrInvalidResponse
	}
	return nil
}

// OwnedLicences returns at most one page. Pass an empty cursor for the first.
func (c *Client) OwnedLicences(ctx context.Context, after string) (OwnedLicences, error) {
	if after != "" && !opaque(after) {
		return OwnedLicences{}, ErrConfiguration
	}
	generation, err := c.generation()
	if err != nil {
		return OwnedLicences{}, err
	}
	if err := c.lockOperation(ctx, generation); err != nil {
		return OwnedLicences{}, err
	}
	defer c.unlockOperation()
	session, err := c.customerSession(generation)
	if err != nil {
		return OwnedLicences{}, err
	}
	data, err := c.transport.GetBearer(ctx, c.accountPath(clientPrefix+"licences", after), session.token)
	var page OwnedLicences
	if err == nil {
		err = decodeJSON(data, &page)
	}
	if err == nil {
		if len(page.Items) > 50 || page.NextCursor != nil && !opaque(*page.NextCursor) {
			err = ErrInvalidResponse
		}
		for _, licence := range page.Items {
			if validation := checkLicence(licence); validation != nil {
				err = validation
				break
			}
		}
	}
	if err := c.finishAccount(ctx, generation, err); err != nil {
		return OwnedLicences{}, err
	}
	return page, nil
}

func (c *Client) ClaimLicence(ctx context.Context, key, idempotencyKey string) (OwnedLicence, error) {
	if key == "" || len(key) > 256 || !validOperationID(idempotencyKey) {
		return OwnedLicence{}, ErrConfiguration
	}
	var licence OwnedLicence
	err := c.accountPost(ctx, clientPrefix+"licence-claims", map[string]any{"licence_key": key, "idempotency_key": idempotencyKey}, true, func(data json.RawMessage) error {
		if err := decodeJSON(data, &licence); err != nil {
			return err
		}
		return checkLicence(licence)
	})
	return licence, err
}
func (c *Client) RequestEmailChange(ctx context.Context, password, email string) error {
	if len(password) > 256 || len(email) > 254 || !utf8.ValidString(password) {
		return ErrConfiguration
	}
	return c.accountPost(ctx, clientPrefix+"email-changes", map[string]any{"password": password, "email": email}, false, checkAccepted)
}
func (c *Client) Register(ctx context.Context, input Registration) (*PendingRegistration, error) {
	if input.LicenceKey == "" || len(input.LicenceKey) > 256 || len(input.Username) > 128 || len(input.Email) > 254 || len(input.Password) > 256 || !utf8.ValidString(input.Password) || utf8.RuneCountInString(input.Password) < 8 {
		return nil, ErrConfiguration
	}
	data, err := c.publicPost(ctx, clientPrefix+"registrations", map[string]any{"licence_key": input.LicenceKey, "username": input.Username, "email": input.Email, "password": input.Password})
	if err != nil {
		return nil, err
	}
	var reply registrationReply
	if decodeJSON(data, &reply) != nil || !reply.Accepted || !bearer(reply.ResendCredential) {
		return nil, ErrInvalidResponse
	}
	if _, err := timestamp(reply.ExpiresAt); err != nil {
		return nil, err
	}
	return &PendingRegistration{Accepted: true, ExpiresAt: reply.ExpiresAt, resendCredential: reply.ResendCredential, applicationID: c.config.ApplicationID, environmentID: c.config.EnvironmentID}, nil
}
func (c *Client) ResendRegistration(ctx context.Context, pending *PendingRegistration) error {
	if pending == nil || pending.applicationID != c.config.ApplicationID || pending.environmentID != c.config.EnvironmentID || !bearer(pending.resendCredential) {
		return ErrConfiguration
	}
	data, err := c.publicPost(ctx, clientPrefix+"registrations/resend", map[string]any{"resend_credential": pending.resendCredential})
	if err != nil {
		return err
	}
	return checkAccepted(data)
}
func (c *Client) RequestPasswordRecovery(ctx context.Context, email string) error {
	if len(email) > 254 {
		return ErrConfiguration
	}
	data, err := c.publicPost(ctx, clientPrefix+"password-recovery", map[string]any{"email": email})
	if err != nil {
		return err
	}
	return checkAccepted(data)
}
func (c *Client) scopeBody(body map[string]any) map[string]any {
	body["application_id"] = c.config.ApplicationID
	body["environment_id"] = c.config.EnvironmentID
	return body
}
func (c *Client) accountPath(route, after string) string {
	result := route + "?application_id=" + c.config.ApplicationID + "&environment_id=" + c.config.EnvironmentID
	if after != "" {
		result += "&after=" + after
	}
	return result
}
func (c *Client) customerSession(generation uint64) (*accountSession, error) {
	c.mu.Lock()
	defer c.mu.Unlock()
	if err := c.syncStorageLocked(); err != nil {
		return nil, err
	}
	if c.state.generation != generation {
		return nil, ErrStaleResponse
	}
	if c.state.account == nil {
		return nil, ErrReauthenticationRequired
	}
	copy := *c.state.account
	return &copy, nil
}
func (c *Client) accountPost(ctx context.Context, route string, body map[string]any, safe bool, validate func(json.RawMessage) error) error {
	generation, err := c.generation()
	if err != nil {
		return err
	}
	if err := c.lockOperation(ctx, generation); err != nil {
		return err
	}
	defer c.unlockOperation()
	session, err := c.customerSession(generation)
	if err != nil {
		return err
	}
	body["customer_session"] = session.token
	data, err := c.transport.Post(ctx, route, c.scopeBody(body), safe)
	if err == nil {
		err = validate(data)
	}
	return c.finishAccount(ctx, generation, err)
}
func (c *Client) publicPost(ctx context.Context, route string, body map[string]any) (json.RawMessage, error) {
	generation, err := c.generation()
	if err != nil {
		return nil, err
	}
	data, responseErr := c.transport.Post(ctx, route, c.scopeBody(body), false)
	if err := c.checkGeneration(generation); err != nil {
		return nil, err
	}
	if ctx.Err() != nil {
		return nil, ErrCancelled
	}
	return data, responseErr
}
func (c *Client) finishAccount(ctx context.Context, generation uint64, err error) error {
	c.mu.Lock()
	defer c.mu.Unlock()
	if syncErr := c.syncStorageLocked(); syncErr != nil {
		return syncErr
	}
	if c.state.generation != generation {
		return ErrStaleResponse
	}
	if ctx.Err() != nil {
		return ErrCancelled
	}
	if err != nil && !errors.Is(err, ErrTransient) && !errors.Is(err, ErrCancelled) && !errors.Is(err, &Error{Kind: Denied, Code: "session_expired"}) {
		clearState(&c.state)
		if storageErr := c.invalidateLocked(); storageErr != nil {
			return storageErr
		}
	}
	return err
}
func checkAccepted(data json.RawMessage) error {
	var reply acceptedReply
	if decodeJSON(data, &reply) != nil || !reply.Accepted {
		return ErrInvalidResponse
	}
	return nil
}
func checkLogin(reply loginReply) error {
	customer := reply.Customer
	if !opaque(customer.ID) || customer.Suspended || len(customer.Username) < 3 || len(customer.Username) > 32 || !errorCode(customer.Username) || len(customer.Email) == 0 || len(customer.Email) > 254 || !bearer(reply.Session) {
		return ErrInvalidResponse
	}
	if _, err := timestamp(customer.CreatedAt); err != nil {
		return err
	}
	if _, err := timestamp(reply.ExpiresAt); err != nil {
		return err
	}
	return nil
}
func checkLicence(licence OwnedLicence) error {
	if !opaque(licence.ID) || licence.DeviceLimit < 1 || licence.DeviceLimit > 100 || !validEntitlements(licence.Entitlements) || utf8.RuneCountInString(licence.PolicyName) > 80 {
		return ErrInvalidResponse
	}
	for _, date := range []*string{licence.FirstUsedAt, licence.ExpiresAt} {
		if date != nil {
			if _, err := timestamp(*date); err != nil {
				return err
			}
		}
	}
	return nil
}
